Module3HM3_Algorithms/SearchingSortingRecursion/CollectionExtension.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SearchingSortingRecursion
+{
+    public static class CollectionExtension
+    {
+        public static int BinarySearch<T>(this IEnumerable<T> collection, T item) where T : IComparable<T>
+        {
+            if (collection == null) throw new ArgumentNullException("collection");
+            if (item == null) throw new ArgumentNullException("item");
+            if (!collection.Any()) return -1;
+
+            IList<T> list = collection is IList<T> ? (IList<T>)collection : new List<T>(collection);
+
+            bool isAscending = true;
+            for (int i = 1; i < list.Count; i++)
```

Comment:
Same, all pre-processing, should not be a part of the algorithm. It should be able to handle unsorted collections in a different manner.